sources/packages/backend/apps/workers/src/services/application/application.service.ts

```diff
@@ -89,4 +109,148 @@ export class ApplicationService extends RecordDataModelService<Application> {
+        `Student Application is not in the expected status. The application must be in application status '${ApplicationStatus.assessment}' for an MSFAA number be assigned.`,
+        INVALID_OPERATION_IN_THE_CURRENT_STATUS,
+      );
+    }
+
+    let msfaaNumberId: number;
+
+    // Checks if there is an MSFAA that could be considered valid.
+    const existingValidMSFAANumber =
+      await this.msfaaNumberService.getCurrentValidMSFAANumber(
+        application.student.id,
+        application.currentAssessment.offering.offeringIntensity,
+      );
+    if (existingValidMSFAANumber) {
+      // Reuse the MSFAA that is still valid and avoid creating a new one.
+      msfaaNumberId = existingValidMSFAANumber.id;
+    } else {
+      // Get previously completed and signed application for the student
+      // to determine if an existing MSFAA is still valid.
+      const previousSignedApplication =
+        await this.getPreviouslySignedApplication(
+          application.student.id,
+          application.currentAssessment.offering.offeringIntensity,
+        );
+
+      let hasValidMSFAANumber = false;
+      if (previousSignedApplication) {
+        const [originalAssessment] =
+          previousSignedApplication.studentAssessments;
+        // checks if the MSFAA number is still valid.
+        hasValidMSFAANumber = this.msfaaNumberService.isMSFAANumberValid(
+          // Previously signed and completed application offering end date in considered the start date.
+          originalAssessment.offering.studyEndDate,
+          // Start date of the offering of the current application is considered the end date.
+          originalAssessment.offering.studyStartDate,
+        );
+      }
+
+      if (hasValidMSFAANumber) {
+        // Reuse the MSFAA number.
+        msfaaNumberId = previousSignedApplication.msfaaNumber.id;
+      } else {
+        // Create a new MSFAA number case the previous one is no longer valid.
+        const newMSFAANumber = await this.msfaaNumberService.createMSFAANumber(
+          application.student.id,
+          applicationId,
+          application.currentAssessment.offering.offeringIntensity,
+        );
+        msfaaNumberId = newMSFAANumber.id;
+      }
+    }
+
+    // Associate the MSFAA number with the application.
+    application.msfaaNumber = { id: msfaaNumberId } as MSFAANumber;
+
+    return this.repo.save(application);
+  }
+
+  /**
+   * Gets the application that has an MSFAA signed date.
+   * @param studentId student id to filter the applications.
+   * @param offeringIntensity MSFAA are generated individually for full-time/part-time
+   * applications. The offering intensity is used to differentiate between them.
+   * @returns previous signed application if exists, otherwise null.
+   */
+  async getPreviouslySignedApplication(
+    studentId: number,
+    offeringIntensity: OfferingIntensity,
+  ): Promise<Application> {
+    return this.repo
+      .createQueryBuilder("applications")
```

Comment:
a general qn for my info, so what decides that we build a query with the orm or with the querybuilder, 
![image](https://user-images.githubusercontent.com/77353155/196787296-90ef471a-2801-4b6f-b11f-73ab7d2ca9c5.png)

Reply:
There is no decision to be made, ALWAYS go for object type query unless there is a really, really good excuse or technical issue 😉 
@dheepak-aot @guru-aot @andrepestana-aot - FYI